output empty creationTimestamp as null

juanvallejo opened the pull request:
**Release note**
```release-note
NONE
```

Updates the value of the `creationTimestamp` field to be `null`
when empty, to keep parity between it and `deletionTimestamp`.

Adds a round-trip test to ensure that unstructured objects containing
empty metadata fields are able to be re-converted back into internal
or external objects. Prior to the proposed patch in this PR, an
unstructured object whose `.metadata.creationTimestamp` value had
been set through the metadata accessor to an empty value 
(`metav1.Time{}` in this case), was unable to be re-converted to an
internal or external type using the runtime decoder. Conversion would
fail with the error:

```
unstructured_test.go:177: FromUnstructured failed: parsing time "" as "2006-01-02T15:04:05Z07:00": cannot parse "" as "2006"
```

cc @liggitt @fabianofranz

juanvallejo committed:
remove empty creationTimestamp field

juanvallejo commented:
/assign fabianofranz
/assign shyiwang

cblecker commented:
/ok-to-test

juanvallejo commented:
/test pull-kubernetes-e2e-gce-bazel

liggitt commented:
Lgtm, would like to see a test

juanvallejo commented:
@liggitt thanks, test added

liggitt commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go:
```diff
@@ -464,6 +464,10 @@ func (u *Unstructured) GetCreationTimestamp() metav1.Time {
 func (u *Unstructured) SetCreationTimestamp(timestamp metav1.Time) {
 	ts, _ := timestamp.MarshalQueryParameter()
+	if len(ts) == 0 {
+		u.setNestedField(nil, "metadata", "creationTimestamp")
```
hmm... actually, should this be storing `nil` or removing from the map? @ironcladlou, any thoughts here?

ironcladlou replied:
I suggested removing nil keys from the map generally in https://github.com/kubernetes/kubernetes/issues/48211, but I wouldn't say there's consensus on that point yet.

@deads2k suggested we rely on `omitempty` in the external types to accommodate a nil here (mirroring his prior counterpoint to https://github.com/kubernetes/kubernetes/issues/48211#issuecomment-311828662), which I'm fine with.

Maybe some more godocs would help here in the meantime?

deads2k replied:
This field is omitempty.

@juanvallejo I prefer to delete the entry, then this lgtm.

juanvallejo replied:
Thanks, will update

juanvallejo commented:
@ironcladlou thanks for your feedback on this; per our conversation I added a round trip test
that ensures empty metadata field values in an object do not inhibit that object's conversion.
Updated PR's description to explain the new test, and did a local test proving the fix from
this patch.

From the PR's description:
```
Adds a round-trip test to ensure that unstructured objects containing
empty metadata fields are able to be re-converted back into internal
or external objects. Prior to the proposed patch in this PR, an
unstructured object whose `.metadata.creationTimestamp` value had
been set through the metadata accessor to an empty value 
(`metav1.Time{}` in this case), was unable to be re-converted to an
internal or external type using the runtime decoder. Conversion would
fail with the error:

unstructured_test.go:177: FromUnstructured failed: parsing time "" as "2006-01-02T15:04:05Z07:00": cannot parse "" as "2006"
```

juanvallejo commented:
/retest

juanvallejo commented:
/retest

juanvallejo commented:
@liggitt @ironcladlou friendly ping

ironcladlou commented on pkg/api/testing/unstructured_test.go:
```diff
@@ -130,6 +133,49 @@ func TestRoundTrip(t *testing.T) {
+			unstructObj.Object = unstrBody
+
+			if meta, err := meta.Accessor(unstructObj); err == nil {
+				meta.SetCreationTimestamp(metav1.Time{})
```
Should the test fail if `err != nil`?

ironcladlou commented on pkg/api/testing/unstructured_test.go:
```diff
@@ -130,6 +133,49 @@ func TestRoundTrip(t *testing.T) {
+
+			_, ok := unstructObj.Object["metadata"]
+			if !ok {
+				t.Logf("No metadata field found for %v, Skipping...", kind)
```
If the item has a meta accessor and the Unstructured is missing a `metadata` key, does that seem like an erroneous condition?

ironcladlou reviewed (commented):
Couple of questions/nits in the unit test, otherwise LGTM.

juanvallejo commented:
/test pull-kubernetes-unit

juanvallejo commented:
@ironcladlou @liggitt thanks for the feedback, wondering if it'd be okay to lgtm this?

cblecker commented:
/test pull-kubernetes-e2e-gce

juanvallejo commented:
ping @ironcladlou or @liggitt

juanvallejo commented:
/assign liggitt
/assign deads2k

juanvallejo commented:
@liggitt friendly ping

deads2k commented:
@sttts this reminds me of something you just did

sttts commented:
@deads2k similar area, not sure it's related https://github.com/kubernetes/kubernetes/pull/54803

juanvallejo commented:
@deads2k updated patch to remove the `setCreationTimestamp` field on an empty value.
Picked the `removeNestedField` helper from https://github.com/kubernetes/kubernetes/pull/51940
PTAL

deads2k commented on staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured/unstructured.go:
```diff
@@ -258,6 +258,10 @@ func (u *Unstructured) GetCreationTimestamp() metav1.Time {
 func (u *Unstructured) SetCreationTimestamp(timestamp metav1.Time) {
 	ts, _ := timestamp.MarshalQueryParameter()
+	if len(ts) == 0 {
+		RemoveNestedField(u.Object, "metadata", "creationTimestamp")
```
@sttts creationtime is an `omitEmpty` field, so when it is zero value it doesn't get serialized.  While I may not detect it this way (@juanvallejo can you check if the passed `timestamp` is zero value), removing the field when this happens is reasonable.

liggitt replied:
actually, since it's a struct, it still does (yay, go json), but logically, should not

juanvallejo replied:
@deads2k 
> @juanvallejo can you check if the passed timestamp is zero value

Done

deads2k commented:
lgtm, please squash for merge.

juanvallejo commented:
@deads2k thanks, commits squashed

juanvallejo commented:
/test pull-kubernetes-e2e-kops-aws

liggitt commented:
/lgtm

fabianofranz commented:
/approve no-issue

fabianofranz commented:
/sig cli
/kind bug
